src/daft-core/src/array/ops/utf8.rs

```diff
@@ -271,6 +270,30 @@ fn replace_on_literal<'a>(
     Ok(Utf8Array::from((name, Box::new(arrow_result?))))
 }
 
+fn substring(s: &str, start: usize, len: Option<usize>) -> Option<&str> {
+    let mut char_indices = s.char_indices().skip(start);
+    let len = match len {
+        Some(len) => {
+            if len == 0 {
+                return None;
+            } else {
+                len
+            }
+        }
+        None => s.len(),
+    };
+
+    if let Some((start_pos, _)) = char_indices.next() {
```

Comment:
Minor nit: You can just do `char_indices.nth(start)` instead of skipping by `start` at the beginning and calling `next` here